src/tools/tidy/src/features/version.rs

```diff
@@ -0,0 +1,82 @@
+use std::str::FromStr;
+use std::num::ParseIntError;
+use std::fmt;
+use std::convert::TryInto;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Version {
+    parts: [u32; 3],
+}
+
+impl fmt::Display for Version {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let x = self.parts.iter().map(|x| x.to_string()).collect::<Vec<_>>().join(".");
+        f.pad(&x)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParseVersionError {
+    ParseIntError(ParseIntError),
+    // core::array::TryFromSlice is not exported from std, so we invent our own variant
+    WrongNumberOfParts
+}
+
+impl From<ParseIntError> for ParseVersionError {
+    fn from(err: ParseIntError) -> Self {
+        ParseVersionError::ParseIntError(err)
+    }
+}
+
+impl FromStr for Version {
+    type Err = ParseVersionError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<_> = s.split('.').map(|part| part.parse()).collect::<Result<_, _>>()?;
```

Comment:
The way this is written, you always allocate. If we write as:
```rust
impl FromStr for Version {
    type Err = ParseVersionError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // Consume three version parts separated by `.`.
        let mut iter = s.split('.').map(|p| Ok(p.parse()?));
        let mut part = || {
            iter.next()
                .unwrap_or(Err(ParseVersionError::WrongNumberOfParts))
        };
        let parts = [part()?, part()?, part()?];

        if let Some(_) = iter.next() {
            // Ensure we don't have more than 3 parts.
            Err(ParseVersionError::WrongNumberOfParts)
        } else {
            Ok(Self { parts })
        }
    }
}
```
we can avoid that.

Reply:
Thanks for the help—fixed!